rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md

```diff
@@ -1 +1,157 @@
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-04-26
+---
+
+# CKB VM Version Selection
+
+## Abstract
+
+This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.
+
+## Motivation
+
+It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.
+
+This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
+
+After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:
+
+* When the `hash_type` is 0, the script group matches code via data hash and will run the code using the CKB VM version 0.
+* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
+* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.
+
+| `hash_type` | matches by       | VM version |
+| ----------- | ---------------- | ---------- |
+| 0           | data hash        | 0          |
+| 1           | type script hash | 1          |
+| 2           | data hash        | 1          |
+
+The VM version 0 deployed with the Lina and Aggron genesis block is version 0.
+
+The VM version 1 will be deployed along ckb2021. The [rfc33] highlights the changes in this version, and [rfc34] lists the new syscalls only available since version 1.
+
+[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md
+[rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md
+
+The transaction is considered invalid if any `hash_type` is not in the allowed values 0, 1, and 2.
+
+See more information about code locating using `hash_type` in [rfc22].
+
+[rfc22]: ../0022-transaction-structure/0022-transaction-structure.md
+
+Because the VM selection algorithm depends on which epoch the transaction belongs to, it is not deterministic for transactions still in the memory pool. The CKB node must run two versions of transaction relay protocols, one for the current CKB version C and another for the next fork version N.
```

Comment:
I feel like this is implementation details. We could add this in an appendix, but not here in the specification.